Userland/Services/ConfigServer/ClientConnection.cpp

```diff
@@ -37,7 +37,17 @@ static Core::ConfigFile& ensure_domain_config(String const& domain)
     VERIFY(!result.is_error());
     watcher_or_error.value()->on_change = [config, domain](auto&) {
         auto new_config = Core::ConfigFile::open(config->filename(), Core::ConfigFile::AllowWriting::Yes);
-        // FIXME: Detect removed keys.
+        for (auto& group : config->groups()) {
+            for (auto& key : config->keys(group)) {
+                if (!new_config->has_key(group, key)) {
+                    for (auto& it : s_connections) {
+                        if (it.value->is_monitoring_domain(domain)) {
+                            it.value->async_notify_removed_key(domain, group, key);
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
Can we refactor this to avoid 2 triple-nested `for` loops in a row? These lists probably aren't so large as to be a big performance concern, but if we move the innermost `for` loop to be the outermost `for` loop, we can combine them a bit, and filter out connections not monitoring the domain earlier:

```c++
watcher_or_error.value()->on_change = [config, domain](auto&) {
    auto new_config = Core::ConfigFile::open(config->filename(), Core::ConfigFile::AllowWriting::Yes);

    for (auto& it : s_connections) {
        if (!it.value->is_monitoring_domain(domain))
            continue;

        for (auto& group : config->groups()) {
            for (auto& key : config->keys(group)) {
                if (!new_config->has_key(group, key))
                    it.value->async_notify_removed_key(domain, group, key);
            }
        }

        // FIXME: Detect type of keys.
        for (auto& group : new_config->groups()) {
            for (auto& key : new_config->keys(group)) {
                auto old_value = config->read_entry(group, key);
                auto new_value = new_config->read_entry(group, key);
                if (old_value != new_value)
                    it.value->async_notify_changed_string_value(domain, group, key, new_value);
            }
        }
    }

    // FIXME: Refactor this whole thing so that we don't need a cache lookup here.
    s_cache.get(domain).value()->config = new_config;
};
```

Reply:
True, but now the more expensive path, "diffing" the keys etc. is now performed for each connection. 🤔
Had the idea to provide a helper like this:
```
+template<typename Callback, typename... Args>
+void for_each_monitoring_connection(String const& domain, ClientConnection* excluded_connection, Callback&& function, Args&&... arguments)
+{
+    for (auto& it : s_connections) {
+        if (it.value->is_monitoring_domain(domain) && (!excluded_connection || it.value != excluded_connection))
+            (it.value->*function)(forward<Args>(arguments)...);
+    }
+}
+
 static Core::ConfigFile& ensure_domain_config(String const& domain)
 {
     auto it = s_cache.find(domain);
@@ -35,16 +44,14 @@ static Core::ConfigFile& ensure_domain_config(String const& domain)
     VERIFY(!watcher_or_error.is_error());
     auto result = watcher_or_error.value()->add_watch(config->filename(), Core::FileWatcherEvent::Type::ContentModified);
     VERIFY(!result.is_error());
+
     watcher_or_error.value()->on_change = [config, domain](auto&) {
         auto new_config = Core::ConfigFile::open(config->filename(), Core::ConfigFile::AllowWriting::Yes);
+
         for (auto& group : config->groups()) {
             for (auto& key : config->keys(group)) {
                 if (!new_config->has_key(group, key)) {
-                    for (auto& it : s_connections) {
-                        if (it.value->is_monitoring_domain(domain)) {
-                            it.value->async_notify_removed_key(domain, group, key);
-                        }
-                    }
+                    for_each_monitoring_connection(domain, nullptr, &ClientConnection::async_notify_removed_key, domain, group, key);
                 }
             }
         }
```

But I am not sure if this is the most readable way of doing it.

Reply:
Ah fair point. We could probably create a list of changed keys, etc. before doing the callback loop, but tbh that's probably addressing a performance problem that might not even exist.

Typically we just pass a lambda to `for_each_*` methods, even if its just a one liner to call another method. Something like:
```c++
for_each_monitoring_connection(domain, [&](auto& connection) {
    connection.async_notify_removed_key(domain, group, key);
});
```

Reply:
Good point, no need to overcomplicate things. I edited the pr with the new method.